src/osdc/Objecter.h

```diff
@@ -1724,6 +1724,8 @@ class Objecter : public md_config_obs_t, public Dispatcher {
   bool keep_balanced_budget = false;
   bool honor_pool_full = true;
 
+  std::atomic<int> extra_read_flags{0};
```

Comment:
Eww do we really need to make this atomic and stick a memory barrier in the middle of all those constructor calls?
All we really need here is for the memory update to not be torn and I don't think that can happen for these sizes?

Reply:
Use the atomic, racing updates from multiple threads are undefined behavior otherwise.

Reply:
if we want to avoid the atomic here, we might consider removing the support for runtime changes. i think we care more about performance here than the convenience around config changes

Reply:
We don't need sequentially consistent memory ordering (default for a C++ atomic), release-acquire ordering should be sufficient - https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering .

On X86 with its strong memory model this results in no memory barriers in the generated code, just constraints on the compiler not to re-order code. On other architectures (e.g. ARM and POWER) release-acquire ordering adds synchronization instructions, but these are still have less overhead than the instructions added for sequential consistent memory order.

So using code like this when reading the flags would be more efficient: extra_read_flags.load(std::memory_order_acquire)

Given that the balanced and localized flags are just performance optimizations, and as such it doesn't really matter if a race hazard between threads mean a few I/Os use an out of date copy of flags, then a weaker memory ordering of memory_order_relaxed could be used

Reply:
Using a weaker memory ordering seems reasonable to me -- my main concern was simply that we use defined C++ memory ordering semantics.